bootstrap/src/main/java/jenkins/bootstrap/Bootstrap.java

```diff
@@ -0,0 +1,313 @@
+ * 
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * 
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ * 
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.bootstrap;
+
+import javax.naming.Context;
+import javax.naming.InitialContext;
+import javax.naming.NamingException;
+import javax.servlet.ServletContext;
+import javax.servlet.ServletContextEvent;
+import javax.servlet.ServletContextListener;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.Set;
+import java.util.logging.Logger;
+
+import static java.util.logging.Level.*;
+
+/**
+ * Entry point when Jenkins is used as a webapp.
+ *
+ * @author Kohsuke Kawaguchi
+ */
+public class Bootstrap implements ServletContextListener {
+
+    private List<BootLogic> bootLogics = Collections.emptyList();
+    private ServletContext context;
+    private ClassLoader coreClassLoader;
+    /**
+     * JENKINS_HOME
+     */
+    private File home;
+
+    /**
+     * Creates the sole instance of {@link jenkins.model.Jenkins} and register it to the {@link ServletContext}.
+     */
+    public void contextInitialized(ServletContextEvent event) {
+        context = event.getServletContext();
+        context.setAttribute(Bootstrap.class.getName(),this);   // allow discovery of this
+        try {
+
+            final FileAndDescription describedHomeDir = getHomeDir(event);
+            home = describedHomeDir.file.getAbsoluteFile();
+            home.mkdirs();
+            System.out.println("Jenkins home directory: "+ home +" found at: "+describedHomeDir.description);
+
+            // check that home exists (as mkdirs could have failed silently), otherwise throw a meaningful error
+            if (!home.exists())
```

Comment:
I classify this as unrelated change, too. The code in question is a copy-paste from `WebAppMain`.